File: photonweave/material_fit.py
"""Constrained passive Drude/Lorentz fitting with independently specified loss.

Linear nonnegative strengths seed a bounded nonlinear fit of oscillator rates.
Positive strengths/damping and epsilon-infinity >= 1 enforce the implemented
isotropic passive model. A finite data band does not validate extrapolation.
"""
from __future__ import annotations

from dataclasses import dataclass
import json
from pathlib import Path
import time
from typing import Literal

import numpy as np
from pydantic import Field, model_validator

from .models import Model, Material, LorentzPole
from .optical_data import OpticalData
from .materials import permittivity


class FitOptions(Model):
    max_poles: int=Field(default=6,ge=1,le=16)
    include_drude: bool=True
    epsilon_inf: float | None=Field(default=None,ge=1,le=400)
    tolerance: float=Field(default=1e-3,gt=0,le=1)
    wavelength_range_um: tuple[float,float] | None=None
    target: Literal['analytic','ade']='analytic'
    dt_s: float=Field(default=0,ge=0)
    loss_weight: float=Field(default=1,gt=0,le=100)
    starts: int=Field(default=2,ge=1,le=6)
    max_nfev: int=Field(default=400,ge=10,le=3000)
    seed: int=Field(default=0,ge=0,le=2**32-1)

    @model_validator(mode='after')
    def valid_options(self):
        if self.target=='ade' and self.dt_s<=0:
            raise ValueError('An ADE-target fit needs a positive simulation timestep.')
        if self.wavelength_range_um is not None and not 0<self.wavelength_range_um[0]<self.wavelength_range_um[1]:
            raise ValueError('Fit wavelength range must be positive and increasing.')
        return self


class OpticalDataRequest(Model):
    text: str=Field(max_length=2_000_000)
    kind: Literal['nk','epsilon']='nk'
    unit: Literal['um','nm','m']='um'
    reference: str=Field(default='',max_length=2000)


class MaterialFitRequest(Model):
    data: OpticalData
    options: FitOptions=Field(default_factory=FitOptions)
    name: str=Field(default='Fitted material',min_length=1,max_length=100)
    color: str='#6ca8dd'


@dataclass
class MaterialFitResult:
    material: Material
    report: dict

    @property
    def converged(self):return self.report['converged']

    def require_tolerance(self):
        if not self.converged:raise ValueError('Material fit did not reach the requested tolerance. Inspect its report.')
        return self.material

    def as_dict(self):return dict(material=self.material.model_dump(),report=self.report)

    def save(self,path):
        Path(path).write_text(json.dumps(self.as_dict(),indent=2)+'\n',encoding='utf8')


def _weights(frequency):
    order=np.argsort(frequency);f=frequency[order]
    width=np.empty(len(f));width[0]=(f[1]-f[0])/2;width[-1]=(f[-1]-f[-2])/2
    width[1:-1]=(f[2:]-f[:-2])/2
    result=np.empty_like(width);result[order]=width/width.sum()
    return result


def _metrics(expected,actual,weights):
    error=abs(actual-expected);scale=np.maximum(abs(expected),1)
    n,k=np.sqrt(expected),np.sqrt(actual)
    return dict(normalized_rms=float(np.sqrt(np.sum(weights*(error/scale)**2))),
                relative_l2=float(np.sqrt(np.sum(weights*error**2)/max(np.sum(weights*abs(expected)**2),1e-300))),
                max_normalized_error=float(np.max(error/scale)),
                max_abs_n_error=float(np.max(abs(n.real-k.real))),
                max_abs_k_error=float(np.max(abs(n.imag-k.imag))))


def material_fit_report(material,*,dt_s=0):
    """Recompute errors for current coefficients, including manual edits."""
    if material.samples is None:raise ValueError('Material has no retained optical samples.')
    data=material.samples;w=np.asarray(data.wavelength_um)
    band=material.fit_band_um or (w[0],w[-1]);mask=(w>=band[0])&(w<=band[1])
    f=data.frequency_hz[mask];measured=data.epsilon[mask];weights=_weights(f)
    analytic=permittivity(material,f);numerical=permittivity(material,f,dt_s) if dt_s else None
    if not np.all(np.isfinite(analytic)) or (numerical is not None and not np.all(np.isfinite(numerical))):
        raise ValueError('A material resonance is singular in the comparison band.')
    index=np.sqrt(measured);fit=np.sqrt(analytic);ade=np.sqrt(numerical) if numerical is not None else None
    return dict(wavelength_um=w[mask].tolist(),measured_n=index.real.tolist(),measured_k=index.imag.tolist(),
                fitted_n=fit.real.tolist(),fitted_k=fit.imag.tolist(),
                numerical_n=None if ade is None else ade.real.tolist(),numerical_k=None if ade is None else ade.imag.tolist(),
                analytic=_metrics(measured,analytic,weights),ade=None if ade is None else _metrics(measured,numerical,weights),
                dt_s=dt_s,fit_band_um=list(band),sample_count=int(mask.sum()),data_sha256=data.fingerprint)


def fit_material(data: OpticalData,*,name='Fitted material',color='#6ca8dd',options: FitOptions | dict | None=None):
    """Fit a passive isotropic material. Returns best fit plus explicit errors.

    The objective integrates squared complex-permittivity error over frequency,
    normalized pointwise by max(1, abs(epsilon)). `max_poles` is an upper bound.
    A tolerance failure is returned explicitly, never converted into success.
    ``target='ade'`` fits the bilinear discrete response at exactly ``dt_s``.
    """
    from scipy.optimize import least_squares, nnls, lsq_linear

    start=time.perf_counter()
    data=OpticalData.model_validate(data.model_dump() if isinstance(data,OpticalData) else data)
    opt=FitOptions.model_validate(options.model_dump() if isinstance(options,FitOptions) else options or {})
    w=np.asarray(data.wavelength_um)
    requested=opt.wavelength_range_um or (float(w[0]),float(w[-1]))
    if requested[0]<w[0] or requested[1]>w[-1]:raise ValueError('Fit range must remain inside the input data band.')
    mask=(w>=requested[0])&(w<=requested[1]);f=data.frequency_hz[mask];target=data.epsilon[mask]
    if len(f)<3:raise ValueError('Fit range needs at least three distinct measured samples.')
    if opt.dt_s and max(f)*opt.dt_s>=.5:raise ValueError('Material frequencies must lie below timestep Nyquist.')
    omega=2*np.pi*f if opt.target=='analytic' else 2/opt.dt_s*np.tan(np.pi*f*opt.dt_s)
    scale=float(np.sqrt(min(omega)*max(omega)));z=omega/scale
    if not np.isfinite(scale) or scale<=0 or max(omega)>=1e18:
        raise ValueError('Optical frequencies exceed the supported oscillator rate range.')
    weights=_weights(f);weighted=np.sqrt(weights)/np.maximum(abs(target),1)
    imag_factor=np.sqrt(opt.loss_weight)
    base=opt.epsilon_inf if opt.epsilon_inf is not None else 1.
    free_constant=opt.epsilon_inf is None
    strength_max=1e40/scale**2
    rlow,rhigh=.1*min(z),min(10*max(z),1e18/scale)
    glow,ghigh=1e-5*min(z),min(20*max(z),1e18/scale)

    def stack(value):
        if value.ndim==1:v=value*weighted
        else:v=value*weighted[:,None]
        return np.concatenate((v.real,imag_factor*v.imag),axis=0)

    def basis(rates):
        return np.column_stack([1/(r*r-z*z-1j*g*z) for r,g in rates]) if rates else np.empty((len(z),0),complex)

    def linear(rates):
        b=basis(rates)
        matrix=np.column_stack((np.ones(len(z)),b)) if free_constant else b
        if matrix.shape[1]==0:return np.empty(0),float(np.dot(stack(base-target),stack(base-target)))
        a=stack(matrix);rhs=stack(target-base);norms=np.linalg.norm(a,axis=0)
        if np.any(norms<=0) or not np.all(np.isfinite(norms)) or not np.all(np.isfinite(rhs)):
            raise ValueError('Optical data magnitude exceeds the numerical range of the fitter.')
        a=a/norms
        try:coeff=nnls(a,rhs,maxiter=max(100,10*a.shape[1]))[0]/norms
        except RuntimeError:coeff=lsq_linear(a,rhs,bounds=(0,np.inf),tol=1e-10).x/norms
        if free_constant and coeff[0]>399:
            rest=nnls(a[:,1:],rhs-399*stack(np.ones(len(z),complex)),maxiter=300)[0]/norms[1:] if rates else np.empty(0)
            coeff=np.r_[399.,rest]
        coeff=np.clip(coeff,0,np.r_[399.,np.full(len(rates),strength_max)] if free_constant else strength_max)
        residual=stack(base+matrix@coeff-target)
        return coeff,float(residual@residual)

    def refine(rates,coeff):
        count=len(rates);nonzero=[i for i,(r,g) in enumerate(rates) if r>0]
        ncoeff=count+int(free_constant)
        x=np.r_[coeff,np.log([rates[i][0] for i in nonzero]),np.log([g for r,g in rates])]
        lower=np.r_[np.zeros(ncoeff),np.full(len(nonzero),np.log(rlow)),np.full(count,np.log(glow))]
        upper=np.r_[([399.] if free_constant else []),np.full(count,strength_max),
                    np.full(len(nonzero),np.log(rhigh)),np.full(count,np.log(ghigh))]

        def model_jac(x):
            c=x[:ncoeff];r=np.zeros(count);r[nonzero]=np.exp(x[ncoeff:ncoeff+len(nonzero)]);g=np.exp(x[-count:])
            a=c[int(free_constant):];den=r[None,:]**2-z[:,None]**2-1j*g[None,:]*z[:,None]
            b=1/den;value=base+(c[0] if free_constant else 0)+b@a
            jac=np.column_stack(([np.ones(len(z))] if free_constant else [])+
                                [b[:,i] for i in range(count)]+
                                [-2*a[i]*r[i]**2/den[:,i]**2 for i in nonzero]+
                                [1j*a[i]*g[i]*z/den[:,i]**2 for i in range(count)])
            return value,jac,r,g

        result=least_squares(lambda x:stack(model_jac(x)[0]-target),np.clip(x,lower,upper),
                             jac=lambda x:stack(model_jac(x)[1]),bounds=(lower,upper),x_scale='jac',
                             max_nfev=opt.max_nfev,ftol=1e-10,xtol=1e-10,gtol=1e-10)
        _,_,r,g=model_jac(result.x)
        return list(zip(r,g)),result.x[:ncoeff],float(result.fun@result.fun),bool(result.success),result.nfev

    def material(rates,coeff):
        eps=base+(float(coeff[0]) if free_constant else 0)
        poles=[LorentzPole(resonance_rad_s=float(r*scale),strength_rad_s_squared=float(a*scale**2),
                           damping_rad_s=float(g*scale))
               for (r,g),a in zip(rates,coeff[int(free_constant):]) if a>1e-14]
        return Material(name=name,color=color,model='multipole' if poles else 'dielectric',epsilon_inf=eps,
                        index=float(np.sqrt(eps)),poles=poles,samples=data,
                        fit_band_um=(float(w[mask][0]),float(w[mask][-1])),fit_dt_s=opt.dt_s if opt.target=='ade' else None)

    rates=[];coeff,cost=linear(rates);history=[];rng=np.random.default_rng(opt.seed)
    best=material(rates,coeff);quality=material_fit_report(best,dt_s=opt.dt_s)
    current_error=quality[opt.target]['normalized_rms']
    history.append(dict(poles=0,normalized_rms=current_error,optimizer_success=True,evaluations=0))
    # Broad deterministic candidates plus measured local loss maxima.
    peaks=[z[i] for i in range(1,len(z)-1) if target.imag[i]>=target.imag[i-1] and target.imag[i]>target.imag[i+1]]
    peaks=sorted(peaks,key=lambda x:float(target.imag[np.argmin(abs(z-x))]),reverse=True)[:12]
    resonance=np.unique(np.r_[np.geomspace(max(rlow,.25*min(z)),min(rhigh,4*max(z)),32),peaks])
    candidates=[(r,g) for r in resonance for g in np.clip(np.array([.01,.05,.2,.7])*r,glow,ghigh)]
    if opt.include_drude:candidates.extend((0.,g) for g in np.geomspace(max(glow,.003*min(z)),min(ghigh,3*max(z)),12))
    pole_limit=min(opt.max_poles,max(1,(2*len(f)-1)//3))
    for count in range(1,pole_limit+1):
        if current_error<=opt.tolerance:break
        trials=[]
        for candidate in candidates:
            if candidate[0]==0 and any(r==0 for r,g in rates):continue
            trial_rates=rates+[candidate];c,score=linear(trial_rates);trials.append((score,trial_rates,c))
        trials.sort(key=lambda t:t[0]);winning=(cost,rates,coeff,False,0);total_nfev=0
        # Zero resonance is a distinct model topology. A bounded nonzero
        # resonance cannot reach it during nonlinear refinement. Comparing
        # only the lowest linear seed scores can therefore exclude Drude
        # even when the measured response is exactly a single Drude pole.
        families=[[t for t in trials if t[1][-1][0]>0],
                  [t for t in trials if t[1][-1][0]==0]]
        for family in families:
            for attempt in range(min(opt.starts,len(family))):
                _,trial_rates,c=family[attempt]
                if attempt:
                    trial_rates=[(0. if r==0 else float(np.clip(r*np.exp(rng.normal(0,.1)),rlow,rhigh)),
                                  float(np.clip(g*np.exp(rng.normal(0,.25)),glow,ghigh))) for r,g in trial_rates]
                    c,_=linear(trial_rates)
                rr,cc,score,success,nfev=refine(trial_rates,c)
                total_nfev+=nfev
                if score<winning[0]:winning=(score,rr,cc,success,nfev)
        cost,rates,coeff,success,nfev=winning
        best=material(rates,coeff);quality=material_fit_report(best,dt_s=opt.dt_s)
        current_error=quality[opt.target]['normalized_rms']
        history.append(dict(poles=len(best.poles),normalized_rms=current_error,optimizer_success=success,evaluations=total_nfev))
        if len(rates)<count:break
    quality.update(converged=current_error<=opt.tolerance,tolerance=opt.tolerance,target=opt.target,
                   pole_count=len(best.poles),requested_max_poles=opt.max_poles,allowed_max_poles=pole_limit,
                   options=opt.model_dump(),history=history,seconds=time.perf_counter()-start,
                   passivity='positive oscillator strengths and damping, epsilon infinity >= 1',
                   error_definition='Frequency-weighted RMS of abs(delta epsilon) / max(1, abs(measured epsilon)).',
                   limitations=['No accuracy claim outside the fitted sample band.',
                                'Bounded nonlinear refinement can converge to a local minimum. Failure to reach tolerance is explicit.',
                                'A passive finite-band fit does not prove consistency or uniqueness of measured data.',
                                'Validate time/mesh convergence and device observables separately.'])
    return MaterialFitResult(best,quality)

File: frontend/src/material_fit.js
import { drawPlot } from './views.js';

export function setupMaterialFit({host,material,editable,api,esc,begin,current,invalidate,timestep,use}) {
 const $=s=>host.querySelector(s);
 const band=material.fit_band_um||[material.samples?.wavelength_um[0]||'',material.samples?.wavelength_um.at(-1)||''];
 host.innerHTML=`<details class="material-fit"><summary>Measured optical data · import and fit</summary>
 <p>Supply your own passive isotropic n/k or complex permittivity samples. The table and its reference are saved with the project.</p>
 <fieldset ${editable?'':'disabled'}>
 <div class="fit-controls"><label>Columns <select aria-label="Optical data columns"><option value="nk">Wavelength, n, k</option><option value="epsilon">Wavelength, ε real, ε imaginary</option></select></label>
 <label>Wavelength unit <select aria-label="Optical wavelength unit"><option value="um">µm</option><option value="nm">nm</option><option value="m">m</option></select></label>
 <label>CSV / text file <input aria-label="Optical data file" type="file" accept=".csv,.txt,.tsv"></label></div>
 <textarea aria-label="Optical data table" rows="5" placeholder="wavelength_um,n,k&#10;1.0,1.50,0.01&#10;1.5,1.49,0.01&#10;2.0,1.48,0.01"></textarea>
 <label class="fit-reference">Data reference <input aria-label="Optical data reference" maxlength="2000" value="${esc(material.samples?.reference||'')}" placeholder="Citation or measurement description"></label>
 <button data-import>Import data</button><span data-data-status>${material.samples?`${material.samples.wavelength_um.length} samples retained`:'No samples imported'}</span>
 <div class="fit-controls"><label>Fit start (µm) <input aria-label="Fit wavelength start" type="number" step="any" min="0" value="${band[0]}"></label>
 <label>Fit stop (µm) <input aria-label="Fit wavelength stop" type="number" step="any" min="0" value="${band[1]}"></label>
 <label>Maximum poles <input aria-label="Maximum fit poles" type="number" min="1" max="16" value="6"></label>
 <label>RMS tolerance <input aria-label="Fit tolerance" type="number" min="0" max="1" step="any" value="0.001"></label>
 <label>Response <select aria-label="Fit response"><option value="analytic">Continuous material</option><option value="ade">FDTD at current timestep</option></select></label>
 <label><input aria-label="Include Drude pole" type="checkbox" checked> Include Drude pole</label></div>
 <button data-fit ${material.samples?'':'disabled'}>Fit optical data</button> <button data-use-fit disabled>Use fitted material</button>
 </fieldset><p class="fit-status" role="status">Fit accuracy applies inside the sampled band. Device accuracy also requires time and mesh convergence.</p>
 <canvas aria-label="Measured and fitted optical response"></canvas></details>`;
 let candidate=null;
 const status=message=>{$('.fit-status').textContent=message;};
 function changed(){candidate=null;$('[data-use-fit]').disabled=true;invalidate();status('Inputs changed. Fit again to update the candidate.');const canvas=$('canvas');canvas.getContext('2d').clearRect(0,0,canvas.width,canvas.height);}
 host.querySelectorAll('input,select,textarea').forEach(input=>input.oninput=changed);
 host.querySelectorAll('textarea,[aria-label="Optical data columns"],[aria-label="Optical wavelength unit"],[aria-label="Optical data file"]').forEach(input=>input.addEventListener('input',()=>{$('[data-fit]').disabled=true;}));
 $('[aria-label="Optical data file"]').onchange=async e=>{
  const token=begin(),file=e.target.files[0];if(!file)return;
  if(file.size>2_000_000){status('Optical data file exceeds 2 MB.');return;}
  try {const text=await file.text();if(!current(token))return;$('textarea').value=text;$('[aria-label="Optical data reference"]').value=file.name;changed();}
  catch(e){if(current(token))status(e.message);}
 };
 $('[data-import]').onclick=async()=>{
  const token=begin();status('Reading optical samples…');
  try {
   const data=await api('/materials/data',{text:$('textarea').value,kind:$('[aria-label="Optical data columns"]').value,
    unit:$('[aria-label="Optical wavelength unit"]').value,reference:$('[aria-label="Optical data reference"]').value});
   if(!current(token))return;
   material.samples=data;material.fit_band_um=null;material.fit_dt_s=null;candidate=null;
   $('[aria-label="Fit wavelength start"]').value=data.wavelength_um[0];$('[aria-label="Fit wavelength stop"]').value=data.wavelength_um.at(-1);
   $('[data-data-status]').textContent=`${data.wavelength_um.length} samples · ${data.wavelength_um[0]}–${data.wavelength_um.at(-1)} µm`;
   $('[data-fit]').disabled=!editable;$('[data-use-fit]').disabled=true;
   invalidate();status('Data imported. Fit to create simulation coefficients.');
  }catch(e){if(current(token))status(e.message);}
 };
 $('[data-fit]').onclick=async()=>{
  const token=begin();candidate=null;$('[data-use-fit]').disabled=true;status('Fitting passive oscillators…');
  try {
   const options={max_poles:Number($('[aria-label="Maximum fit poles"]').value),tolerance:Number($('[aria-label="Fit tolerance"]').value),
    wavelength_range_um:[Number($('[aria-label="Fit wavelength start"]').value),Number($('[aria-label="Fit wavelength stop"]').value)],
    include_drude:$('[aria-label="Include Drude pole"]').checked,target:$('[aria-label="Fit response"]').value};
   const data=structuredClone(material.samples);data.reference=$('[aria-label="Optical data reference"]').value;
   options.dt_s=await timestep();if(!current(token))return;
   const result=await api('/materials/fit',{data,options,name:material.name,color:material.color});if(!current(token))return;
   candidate=result;const r=result.report,key=r.target==='ade'?'numerical':'fitted';
   drawPlot($('canvas'),[['measured_n','n (data)'],['measured_k','k (data)'],[`${key}_n`,'n (fit)'],[`${key}_k`,'k (fit)']].map(([key,name])=>({name,wavelength_um:r.wavelength_um,spectrum:r[key],spectrum_label:'Wavelength (µm) · measured and fitted n + i k'})),true,true);
   status(`${r.converged?'Tolerance met':'Tolerance NOT met'} · ${r.pole_count} poles · ${r.sample_count} samples · analytic RMS ${r.analytic.normalized_rms.toExponential(3)} · FDTD RMS ${r.ade.normalized_rms.toExponential(3)} at Δt ${(r.dt_s*1e15).toPrecision(5)} fs · ${r.seconds.toFixed(2)} s. ${r.converged?'Use fitted material, then Apply materials to save.':'Adjust the fit band or pole limit. Current coefficients have not changed.'}`);
   $('[data-use-fit]').disabled=!editable||!r.converged;
  }catch(e){if(current(token))status(e.message);}
 };
 $('[data-use-fit]').onclick=()=>{if(candidate?.report.converged)use(candidate.material);};
 return {invalidate(){candidate=null;$('[data-use-fit]').disabled=true;status('Material parameters changed. Fit again to update the candidate.');}};
}
